In [5]:
import pandas as pd
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', None)
pd.options.mode.chained_assignment = None 

In [6]:
# dfCim = pd.read_csv("https://data.bts.gov/resource/5rpz-kgm9.csv")

## Analysis of Original vs New Source

### Read Old Data and New Source

In [7]:
dfNew= pd.read_csv("PortDataExport-New.csv")
dfOld= pd.read_csv("Port_Data-Old.csv")

In [8]:
print(dfOld.shape)
print(dfNew.shape)

(3637, 11)
(5557, 11)


#### Do the Columns Match

In [178]:
sorted(dfOld.columns)

['Cargo Type',
 'Percent Change',
 'Port ID',
 'Port Ranking',
 'Port_Name',
 'Region',
 'Reporting Year',
 'State',
 'Trade Type',
 'Units',
 'Volume']

In [179]:
sorted(dfNew.columns)

['Cargo Type',
 'Percent Change',
 'Port ID',
 'Port Ranking',
 'Port_Name',
 'Region',
 'Reporting Year',
 'State',
 'Trade Type',
 'Units',
 'Volume']

##### Conclusion
<div style="border: 1px solid blue; padding: 10px;"><span style="font-size: 14pt; color: #169179;">
Yes... Columns are the Same!!
    </div>  

In [180]:
dfNew["Port_Name"].value_counts()


Port Freeport, TX                         159
Mobile, AL                                159
Jacksonville, FL                          159
New York, NY & NJ                         159
Los Angeles, CA Port of                   159
San Juan, PR                              159
Honolulu, O'ahu, HI                       159
Philadelphia Regional Port, PA            159
Houston Port Authority, TX                159
Port of Charleston, SC, SC                159
Virginia, VA, Port of                     159
Long Beach, CA Port of                    159
Baltimore, MD                             159
Savannah, GA Port of                      159
Boston, MA                                158
Port Everglades, FL                       155
Alaska, AK Port of                        155
Wilmington, NC                            152
Port of Portland, OR                      147
Tampa Port Authority, FL                  135
New Orleans, LA                           135
Tacoma, WA                        

#### Month-Year Dates in New Data

In [9]:
dfNew["Reporting Year"].value_counts()

2021       1421
2020        582
2019        549
2018        539
2017        539
3-2020       81
8-2020       81
2-2020       81
1-2021       81
1-2020       81
2-2021       81
7-2020       81
11-2020      81
6-2020       81
10-2020      81
5-2020       81
12-2020      81
9-2021       80
8-2021       80
12-2021      80
9-2020       80
11-2021      80
5-2021       80
10-2021      80
3-2021       80
7-2021       80
4-2021       80
6-2021       78
4-2020       77
Name: Reporting Year, dtype: int64

In [10]:
a =dfNew.loc[dfNew["Reporting Year"].str.len() > 4]

In [11]:
print(a["Cargo Type"].value_counts())
print(a["Reporting Year"].value_counts())

VESSEL DWELL TIMES    1927
Name: Cargo Type, dtype: int64
6-2020     81
5-2020     81
7-2020     81
2-2021     81
11-2020    81
8-2020     81
2-2020     81
10-2020    81
12-2020    81
1-2020     81
3-2020     81
1-2021     81
3-2021     80
11-2021    80
9-2020     80
7-2021     80
9-2021     80
12-2021    80
10-2021    80
8-2021     80
5-2021     80
4-2021     80
6-2021     78
4-2020     77
Name: Reporting Year, dtype: int64


#### Dwell Times in New Data

In [12]:
print("OLD")
print(dfOld["Cargo Type"].value_counts())
print("-----")
print("New")
print(dfNew["Cargo Type"].value_counts())

OLD
TOTAL TONNAGE                  1150
DRY BULK                       1140
CONTAINER                       528
TOP 5 COMMODITIES               297
TOP 5 FOOD/FARM COMMODITIES     272
VESSEL CALLS                    250
Name: Cargo Type, dtype: int64
-----
New
VESSEL DWELL TIMES             1927
TOTAL TONNAGE                  1135
DRY BULK                       1130
CONTAINER                       580
TOP 5 COMMODITIES               286
TOP 5 FOOD/FARM COMMODITIES     259
VESSEL CALLS                    240
Name: Cargo Type, dtype: int64


In [13]:
dfNew.loc[dfNew["Cargo Type"] == "VESSEL DWELL TIMES","Reporting Year"].value_counts()

6-2020     81
5-2020     81
7-2020     81
2-2021     81
11-2020    81
8-2020     81
2-2020     81
10-2020    81
12-2020    81
1-2020     81
3-2020     81
1-2021     81
3-2021     80
11-2021    80
9-2020     80
7-2021     80
9-2021     80
12-2021    80
10-2021    80
8-2021     80
5-2021     80
4-2021     80
6-2021     78
4-2020     77
Name: Reporting Year, dtype: int64

##### Conclusion
<div style="border: 1px solid blue; padding: 10px;"><span style="font-size: 14pt; color: #169179;">

The <b>New data</b> has a Cargo Type, Vessel Dwell Times, that is not available in the Old data, and these have a Reporting Time of <b>Month-Year</b>, as opposed to only Year for all other Cargo Types, in both the Old and New data.

The Dwell Times only occur in 2020 and 2021
    </div>


Remove Vessel Dwell Times from new data

In [14]:
dfNewNoDwell = dfNew.loc[dfNew["Cargo Type"] != "VESSEL DWELL TIMES"]
print(dfNew.shape)
print(dfNewNoDwell.shape)
print(dfNew.shape[0]-dfNewNoDwell.shape[0])

(5557, 11)
(3630, 11)
1927


In [15]:
print(dfOld.shape)
print(dfNewNoDwell.shape)

(3637, 11)
(3630, 11)


#### Compare Available Years

In [16]:
print(dfOld["Reporting Year"].value_counts())
print(dfNewNoDwell["Reporting Year"].value_counts())

2020    1431
2019     559
2017     549
2018     549
2016     549
Name: Reporting Year, dtype: int64
2021    1421
2020     582
2019     549
2017     539
2018     539
Name: Reporting Year, dtype: int64


##### Conclusion
<div style="border: 1px solid blue; padding: 10px;"><span style="font-size: 14pt; color: #169179;">
New Data does NOT have any 2016 Data

New Data DOES have 2021 Data
    </div>

In [17]:
dfOldNo2016 = dfOld.loc[dfOld["Reporting Year"].astype(int) > 2016]
dfNewNoDwellNo2021 = dfNewNoDwell.loc[dfNewNoDwell["Reporting Year"].astype(int) < 2021]

In [18]:
print(dfOldNo2016["Reporting Year"].value_counts())
print(dfNewNoDwellNo2021["Reporting Year"].value_counts())

2020    1431
2019     559
2017     549
2018     549
Name: Reporting Year, dtype: int64
2020    582
2019    549
2017    539
2018    539
Name: Reporting Year, dtype: int64


#### Why the Large Record Count Difference in 2020

In [19]:
print(dfOldNo2016.loc[dfOldNo2016["Reporting Year"].astype(int) == 2020,"Cargo Type"].value_counts())

TOP 5 COMMODITIES              297
TOP 5 FOOD/FARM COMMODITIES    272
TOTAL TONNAGE                  250
DRY BULK                       250
VESSEL CALLS                   250
CONTAINER                      112
Name: Cargo Type, dtype: int64


In [20]:
print(dfNewNoDwellNo2021.loc[dfNewNoDwellNo2021["Reporting Year"].astype(int) == 2020,"Cargo Type"].value_counts())

TOTAL TONNAGE    235
DRY BULK         235
CONTAINER        112
Name: Cargo Type, dtype: int64


In [21]:
dfOldNo2016.loc[(dfOldNo2016["Reporting Year"].astype(int) == 2020) & ~(dfOldNo2016["Cargo Type"].isin(["DRY BULK","TOTAL TONNAGE","CONTAINER"])),"Cargo Type"].value_counts()

TOP 5 COMMODITIES              297
TOP 5 FOOD/FARM COMMODITIES    272
VESSEL CALLS                   250
Name: Cargo Type, dtype: int64

In [22]:
print(dfNew.loc[dfNew["Reporting Year"].str[-4:].astype(int) == 2021,"Cargo Type"].value_counts())

VESSEL DWELL TIMES             960
TOP 5 COMMODITIES              286
TOP 5 FOOD/FARM COMMODITIES    259
TOTAL TONNAGE                  240
DRY BULK                       240
VESSEL CALLS                   240
CONTAINER                      156
Name: Cargo Type, dtype: int64


##### Conclusion
<div style="border: 1px solid blue; padding: 10px;"><span style="font-size: 14pt; color: #169179;">
<pre>
The Old Data, in year 2020,  has Cargo Type: 
TOP 5 COMMODITIES              297
TOP 5 FOOD/FARM COMMODITIES    272
VESSEL CALLS                   250
</pre>
The New data does not have these Cargo Types in 2020, Howeever, they ARE in 2021
    </div>

In [23]:
dfNewNoDwellNo2021["Cargo Type"].value_counts()

TOTAL TONNAGE    895
DRY BULK         890
CONTAINER        424
Name: Cargo Type, dtype: int64

In [24]:
dfOldNo2016["Cargo Type"].value_counts()

TOTAL TONNAGE                  925
DRY BULK                       920
CONTAINER                      424
TOP 5 COMMODITIES              297
TOP 5 FOOD/FARM COMMODITIES    272
VESSEL CALLS                   250
Name: Cargo Type, dtype: int64

In [25]:
dfOldNo2016NoExtra = dfOldNo2016.loc[dfOldNo2016["Cargo Type"].isin(["TOTAL TONNAGE","DRY BULK","CONTAINER"])]
dfOldNo2016NoExtra.shape

(2269, 11)

In [26]:
dfNewNoDwellNo2021.shape

(2209, 11)

#### SUMMARY
<div style="border: 1px solid blue; padding: 10px;"><span style="font-size: 14pt; color: #169179;">

So, to get an Old and New Version of the Data I could compare, I had to remove thes thing: <p>
OLD:<br>
Year 2016<br>
Cargo Types in 2020 :  TOP 5 COMMODITIES,TOP 5 FOOD/FARM COMMODITIES, VESSEL CALLS<br>
</p>
<p>
NEW:<br>
2021
Cargo Types in 2020,2021: VESSEL DWELL TIMES
    </div>

#### Compare Data for 2017-2020

In [199]:
#dfOldNo2016NoExtra
#dfNewNoDwellNo2021

In [ ]:
for year in dfOldNo2016NoExtra["Reporting Year"].unique().tolist():
    tmpOld = dfOldNo2016NoExtra.loc[dfOldNo2016NoExtra["Reporting Year"] == year]
    tmpNew = dfNewNoDwellNo2021.loc[dfNewNoDwellNo2021["Reporting Year"].astype(int) == year]
    string="Cargo Type"
    print("Year ",year)
#    print("Old\n",tmpOld[string].value_counts())
#    print("New\n",tmpNew[string].value_counts())
    
    for port in tmpOld["Port_Name"].unique().tolist():
        tO = tmpOld.loc[tmpOld["Port_Name"] == port]
        tN = tmpNew.loc[tmpNew["Port_Name"] == port]
        print(year,port,tO.shape[0],tN.shape[0])
   #     tN.drop("Report Year New",inplace=True,axis=1)
        tO.reset_index(inplace=True,drop=True)
        tN.reset_index(inplace=True,drop=True)
        
#         print(tO.columns)
#         print(tN.columns)
        tN["Reporting Year"] = tN["Reporting Year"].astype(int)
        col='Percent Change'
        tN[col]= tN[col].str.replace(",","")
        tN[col] = tN[col].astype(float)
        tO[col] = tO[col].apply(lambda x: '{:,.2f}'.format(x))
        tN[col] = tN[col].apply(lambda x: '{:,.2f}'.format(x))
        col='Volume'
        tN[col] = tN[col].str.replace(",","")
        tN[col] = tN[col].astype(float)
#         print("-------------------\nOLD")
#         print(tO)
#         print("-------------------\nNEW")
        
#         print(tN)
        
        # tO[col] = tO[col].apply(lambda x: '{:,.2f}'.format(x))
        # tN[col] = tN[col].apply(lambda x: '{:,.2f}'.format(x))
        if tN.shape[0] > 0:
            a = tO.compare(tN)
            if len(a) > 0:
                print(tO.compare(tN))
        

#### New Way

-------------------------------------------------------------------------------------

##### OLD vs NEW

In [ ]:
df1 = dfOldNo2016NoExtra
df2 = dfNewNoDwellNo2021
stats = {}
newPorts = {}

bad = {}
for port in df1["Port_Name"].unique().tolist():
    tmp1 = df1.loc[df1["Port_Name"] == port]
    tmp2 = df2.loc[df2["Port_Name"] == port]
    stats[port] = {}
    
#    for year in tmp1["Reporting Year"].unique().tolist():
    for year in [2017,2018,2019,2020]:
        
        tO = tmp1.loc[tmp1["Reporting Year"].astype(int) == year]
        tN = tmp2.loc[tmp2["Reporting Year"].astype(int) == year]
        if tO.shape[0] > 0:
            print(port,year,tO.shape[0],tN.shape[0])
            tO.reset_index(inplace=True,drop=True)
            tN.reset_index(inplace=True,drop=True)

    #         print(tO.columns)
    #         print(tN.columns)
            tN["Reporting Year"] = tN["Reporting Year"].astype(int)
            col='Percent Change'
            tN[col]= tN[col].str.replace(",","")
            tN[col] = tN[col].astype(float)
            tO[col] = tO[col].apply(lambda x: '{:,.2f}'.format(x))
            tN[col] = tN[col].apply(lambda x: '{:,.2f}'.format(x))
            col='Volume'
            tN[col] = tN[col].str.replace(",","")
            tN[col] = tN[col].astype(float)
    #         print("-------------------\nOLD")
    #         print(tO)
    #         print("-------------------\nNEW")

    #         print(tN)

            # tO[col] = tO[col].apply(lambda x: '{:,.2f}'.format(x))
            # tN[col] = tN[col].apply(lambda x: '{:,.2f}'.format(x))
            if tN.shape[0] > 0:
                a = tO.compare(tN)
                if len(a) > 0:
                    print(tO.compare(tN))
                    stats[port][year]="F"
                    if port not in bad:
                        bad[port] = {}
                    if year not in bad[port]:
                        bad[port][year]=[]
                    bad[port][year].append(a)
                else:
                    stats[port][year]="P"
            else:
                stats[port][year]="N"
                if port not in newPorts:
                    newPorts[port]={}
                newPorts[port][year]=1
        else:
                if tN.shape[0] == 0:
                    stats[port][year]=" "
                else:
                    stats[port][year]="O"
            
                
        

In [202]:
ports = []
s2017 = []
s2018 = []
s2019 = []
s2020 = []

for port,anl in stats.items():
    print(port,anl)
    ports.append(port)
    # for year in anl:
    #     print(year,type(year))
    s2017.append(anl[2017])
    s2018.append(anl[2018])
    s2019.append(anl[2019])
    s2020.append(anl[2020])
    
                
        

Boston, MA {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
New York, NY & NJ {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
South Jersey Port, NJ {2017: ' ', 2018: ' ', 2019: ' ', 2020: 'P'}
Philadelphia Regional Port, PA {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Wilmington, DE {2017: 'F', 2018: 'F', 2019: 'F', 2020: 'F'}
Baltimore, MD {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Wilmington, NC {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Charleston, SC Port of {2017: 'N', 2018: 'N', 2019: 'N', 2020: 'N'}
Savannah, GA Port of {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Port Everglades, FL {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
San Juan, PR {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Mobile, AL {2017: 'F', 2018: 'F', 2019: 'F', 2020: 'F'}
Jacksonville, FL {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Tampa Port Authority, FL {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Houston Port Authority, TX {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Gulfport, MS {2017: 'N', 2018: 'N', 2019: 

In [203]:
dfStats = pd.DataFrame({
    "Port":ports,
    "2017":s2017,
    "2018":s2018,
    "2019":s2019,
    "2020":s2020
})

In [204]:
dfStats

,Port,2017,2018,2019,2020
0,"Boston, MA",P,P,P,P
1,"New York, NY & NJ",P,P,P,P
2,"South Jersey Port, NJ",,,,P
3,"Philadelphia Regional Port, PA",P,P,P,P
4,"Wilmington, DE",F,F,F,F
5,"Baltimore, MD",P,P,P,P
6,"Wilmington, NC",P,P,P,P
7,"Charleston, SC Port of",N,N,N,N
8,"Savannah, GA Port of",P,P,P,P
9,"Port Everglades, FL",P,P,P,P


In [205]:
print(" OLD vs NEW")
for port,anl in bad.items():
  #  print(port)
    years = list(anl.keys())
    year = years[0]
    a=anl[year][0]
    print(f"{port:25s}  {a.columns.to_list()[0][0]}  OLD: {a.iloc[0,[0]].values.tolist()[0]}  NEW: {a.iloc[0,[1]].values.tolist()[0]}")
#    print("------------------------------------------------------------------------------------")



 OLD vs NEW
Wilmington, DE             Port ID  OLD: 554  NEW: 555
Mobile, AL                 Port ID  OLD: 2005  NEW: 2032
PortMiami, FL              Port ID  OLD: 2164  NEW: 1992
Honolulu, O'ahu, HI        Port ID  OLD: 4420  NEW: 4421
Tacoma, WA                 Port ID  OLD: 4720  NEW: 4719
New Bourbon Port, MO       Region  OLD: nan  NEW: Gulf Coast & Mississippi River
Corpus Christi, TX         Port ID  OLD: 2423  NEW: 2436
Northern Indiana, IN       Region  OLD: nan  NEW: Great Lakes


In [206]:
ports = []
years = {}
for year in [2017,2018,2019,2020]:
    years[year]=[]
for port,anl in newPorts.items():
    ports.append(port)
    
    for year in [2017,2018,2019,2020]:
        if year in anl:
            flag="Y"
        else:
            flag=""
        years[year].append(flag)

In [207]:
dfNewPort = pd.DataFrame({
    "Port":ports,
    "2017":years[2017],
    "2018":years[2018],
    "2019":years[2019],
    "2020":years[2020]
})

In [208]:
display(dfNewPort)

,Port,2017,2018,2019,2020
0,"Charleston, SC Port of",Y,Y,Y,Y
1,"Gulfport, MS",Y,Y,Y,Y
2,"Portland, OR Port of",Y,Y,Y,Y
3,"Cincinnati-Northern KY, Ports of",Y,Y,Y,Y
4,"Huntington-Tristate, KY, OH, WV",Y,Y,Y,Y
5,"Mid-Ohio Valley Port, OH and WV",,,,Y
6,"St. Louis, MO and IL",Y,Y,Y,Y
7,Illinois Waterway Ports and Terminals Port Sta...,,,,Y


-------------------------------------------------------------------------------------------

##### NEW vs OLD

In [ ]:
#df1 = dfOldNo2016NoExtra
#df2 = dfNewNoDwellNo2021
df2 = dfOldNo2016NoExtra
df1 = dfNewNoDwellNo2021

df1["Reporting Year"]=df1["Reporting Year"].astype(int)
df2["Reporting Year"]=df2["Reporting Year"].astype(int)

stats2 = {}
bad2 = {}
newPorts2 = {}

for port in df1["Port_Name"].unique().tolist():
    tmp1 = df1.loc[df1["Port_Name"] == port]
    tmp2 = df2.loc[df2["Port_Name"] == port]
    stats2[port] = {}
    
#    for year in tmp1["Reporting Year"].unique().tolist():
    for year in [2017,2018,2019,2020]:
        
        tO = tmp1.loc[tmp1["Reporting Year"].astype(int) == year]
        tN = tmp2.loc[tmp2["Reporting Year"].astype(int) == year]
        if tO.shape[0] > 0:
            print(port,year,tO.shape[0],tN.shape[0])
            tO.reset_index(inplace=True,drop=True)
            tN.reset_index(inplace=True,drop=True)

    #         print(tO.columns)
    #         print(tN.columns)
            tN["Reporting Year"] = tN["Reporting Year"].astype(int)
            col='Percent Change'
            tO[col]= tO[col].str.replace(",","")
            tO[col] = tO[col].astype(float)
            tN[col] = tN[col].apply(lambda x: '{:,.2f}'.format(x))
            tO[col] = tO[col].apply(lambda x: '{:,.2f}'.format(x))
            col='Volume'
            tO[col] = tO[col].str.replace(",","")
            tO[col] = tO[col].astype(float)

            if tN.shape[0] > 0:
                a = tO.compare(tN)
                if len(a) > 0:
                    print(tO.compare(tN))
                    stats2[port][year]="F"
                    if port not in bad2:
                        bad2[port] = {}
                    if year not in bad2[port]:
                        bad2[port][year]=[]
                    bad2[port][year].append(a)
                else:
                    stats2[port][year]="P"
            else:
                stats2[port][year]="N"
                if port not in newPorts2:
                    newPorts2[port]={}
                newPorts2[port][year]=1
        else:
                if tN.shape[0] == 0:
                    stats2[port][year]=" "
                else:
                    stats2[port][year]="O"
            
                
        

In [210]:
ports2 = []
s20172 = []
s20182 = []
s20192 = []
s20202 = []

for port,anl in stats2.items():
    print(port,anl)
    ports2.append(port)
    # for year in anl:
    #     print(year,type(year))
    s20172.append(anl[2017])
    s20182.append(anl[2018])
    s20192.append(anl[2019])
    s20202.append(anl[2020])
    


dfStats2 = pd.DataFrame({
    "Port":ports2,
    "2017":s20172,
    "2018":s20182,
    "2019":s20192,
    "2020":s20202
})

Boston, MA {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
New York, NY & NJ {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
South Jersey Port, NJ {2017: ' ', 2018: ' ', 2019: ' ', 2020: 'P'}
Philadelphia Regional Port, PA {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Wilmington, DE {2017: 'F', 2018: 'F', 2019: 'F', 2020: 'F'}
Baltimore, MD {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Wilmington, NC {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Port of Charleston, SC, SC {2017: 'N', 2018: 'N', 2019: 'N', 2020: 'N'}
Savannah, GA Port of {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Port Everglades, FL {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
San Juan, PR {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Port of Palm Beach District, FL {2017: 'N', 2018: 'N', 2019: 'N', 2020: 'N'}
PortMiami, FL {2017: 'F', 2018: 'F', 2019: 'F', 2020: 'F'}
Jacksonville, FL {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Tampa Port Authority, FL {2017: 'P', 2018: 'P', 2019: 'P', 2020: 'P'}
Houston Port Authority, TX {20

In [211]:
display(dfStats2)

,Port,2017,2018,2019,2020
0,"Boston, MA",P,P,P,P
1,"New York, NY & NJ",P,P,P,P
2,"South Jersey Port, NJ",,,,P
3,"Philadelphia Regional Port, PA",P,P,P,P
4,"Wilmington, DE",F,F,F,F
5,"Baltimore, MD",P,P,P,P
6,"Wilmington, NC",P,P,P,P
7,"Port of Charleston, SC, SC",N,N,N,N
8,"Savannah, GA Port of",P,P,P,P
9,"Port Everglades, FL",P,P,P,P


In [212]:
print(" NEW vs OLD")
for port,anl in bad2.items():
  #  print(port)
    years = list(anl.keys())
    year = years[0]
    a=anl[year][0]
    print(f"{port:25s}  {a.columns.to_list()[0][0]}  OLD: {a.iloc[0,[0]].values.tolist()[0]}  NEW: {a.iloc[0,[1]].values.tolist()[0]}")
#    print("------------------------------------------------------------------------------------")

 NEW vs OLD
Wilmington, DE             Port ID  OLD: 555  NEW: 554
PortMiami, FL              Port ID  OLD: 1992  NEW: 2164
Mobile, AL                 Port ID  OLD: 2032  NEW: 2005
Honolulu, O'ahu, HI        Port ID  OLD: 4421  NEW: 4420
Tacoma, WA                 Port ID  OLD: 4719  NEW: 4720
New Bourbon Port, MO       Region  OLD: Gulf Coast & Mississippi River  NEW: nan
Corpus Christi, TX         Port ID  OLD: 2436  NEW: 2423
Northern Indiana, IN       Region  OLD: Great Lakes  NEW: nan


In [213]:
print(" OLD vs NEW")
for port,anl in bad.items():
  #  print(port)
    years = list(anl.keys())
    year = years[0]
    a=anl[year][0]
    print(f"{port:25s}  {a.columns.to_list()[0][0]}  OLD: {a.iloc[0,[0]].values.tolist()[0]}  NEW: {a.iloc[0,[1]].values.tolist()[0]}")

 OLD vs NEW
Wilmington, DE             Port ID  OLD: 554  NEW: 555
Mobile, AL                 Port ID  OLD: 2005  NEW: 2032
PortMiami, FL              Port ID  OLD: 2164  NEW: 1992
Honolulu, O'ahu, HI        Port ID  OLD: 4420  NEW: 4421
Tacoma, WA                 Port ID  OLD: 4720  NEW: 4719
New Bourbon Port, MO       Region  OLD: nan  NEW: Gulf Coast & Mississippi River
Corpus Christi, TX         Port ID  OLD: 2423  NEW: 2436
Northern Indiana, IN       Region  OLD: nan  NEW: Great Lakes


In [214]:
ports2 = []
years2 = {}
for year in [2017,2018,2019,2020]:
    years2[year]=[]
for port,anl in newPorts2.items():
    ports2.append(port)
    
    for year in [2017,2018,2019,2020]:
        if year in anl:
            flag="Y"
        else:
            flag=""
        years2[year].append(flag)
dfNewPort2 = pd.DataFrame({
    "Port":ports2,
    "2017":years2[2017],
    "2018":years2[2018],
    "2019":years2[2019],
    "2020":years2[2020]
})
display(dfNewPort2)


,Port,2017,2018,2019,2020
0,"Port of Charleston, SC, SC",Y,Y,Y,Y
1,"Port of Palm Beach District, FL",Y,Y,Y,Y
2,"Port of Portland, OR",Y,Y,Y,Y
3,"Toledo-Lucas County Port, OH",Y,Y,Y,Y
4,"Cleveland-Cuyahoga Port, OH",Y,Y,Y,Y


In [215]:
display(dfNewPort)

,Port,2017,2018,2019,2020
0,"Charleston, SC Port of",Y,Y,Y,Y
1,"Gulfport, MS",Y,Y,Y,Y
2,"Portland, OR Port of",Y,Y,Y,Y
3,"Cincinnati-Northern KY, Ports of",Y,Y,Y,Y
4,"Huntington-Tristate, KY, OH, WV",Y,Y,Y,Y
5,"Mid-Ohio Valley Port, OH and WV",,,,Y
6,"St. Louis, MO and IL",Y,Y,Y,Y
7,Illinois Waterway Ports and Terminals Port Sta...,,,,Y


##### Conclusion

<div style="border: 1px solid blue; padding: 10px;"><span style="font-size: 14pt; color: #169179;">

~40 Stations in Both Datasets<br>
 For 2017->2020, Most of the Data is the same<br>
<p> 
Exceptions:<br>

5 Ports in New not in Original<br>
8 Ports in Original not in New<br>

6 Ports have Different Port IDs<br>
2 Have regions added in New data<br>
    </div>

----------------------------------------------------------------------------------------------------------

In [216]:
dfOldNo2016NoExtra.columns

Index(['Cargo Type', 'Port ID', 'Port_Name', 'Region', 'Reporting Year',
       'State', 'Trade Type', 'Units', 'Port Ranking', 'Percent Change',
       'Volume'],
      dtype='object')

##### Checks

In [217]:
a = dfOldNo2016NoExtra["Port_Name"].unique().tolist()
b = dfNewNoDwellNo2021["Port_Name"].unique().tolist()

In [218]:
print(len(a),len(b))

50 47


In [219]:
set(a)-set(b)

{'Charleston, SC Port of',
 'Cincinnati-Northern KY, Ports of',
 'Gulfport, MS',
 'Huntington-Tristate, KY, OH, WV',
 'Illinois Waterway Ports and Terminals Port Statistical Area, IL',
 'Mid-Ohio Valley Port, OH and WV',
 'Portland, OR Port of',
 'St. Louis, MO and IL'}

In [220]:
set(b)-set(a)

{'Cleveland-Cuyahoga Port, OH',
 'Port of Charleston, SC, SC',
 'Port of Palm Beach District, FL',
 'Port of Portland, OR',
 'Toledo-Lucas County Port, OH'}

## Create Final Dataset

### Read Original and New data

In [27]:
dfNew= pd.read_csv("PortDataExport-New.csv")
dfOld= pd.read_csv("Port_Data-Old.csv")

In [28]:
dfOld["Reporting Year"].value_counts()

2020    1431
2019     559
2017     549
2018     549
2016     549
Name: Reporting Year, dtype: int64

In [29]:
#  set reporting year to be a string in Original source to match New source
dfOld["Reporting Year"]=dfOld["Reporting Year"].astype(str)

In [30]:
# Seed teh final dataframe with a copy of the New dataframe
dfFinal = dfNew.copy()

In [31]:
print(dfOld.shape)
print(dfFinal.shape)
print(dfNew.shape)

(3637, 11)
(5557, 11)
(5557, 11)


### Add 2016

The New data source starts in 2017, so we need to add teh 2016 data from the Original Source to the New Source

In [32]:
## Create a dataframe with JUST teh 2016 data in it that can be appended to the Final Dataframe
dfOld2016 = dfOld.loc[dfOld["Reporting Year"] == '2016']
print(dfOld2016.shape)

(549, 11)


In [33]:
# Add 2016 to the FInal data frame
dfFinal = pd.concat([dfOld2016,dfFinal])

In [34]:
dfFinal.shape

(6106, 11)

In [35]:
# Check to see if 2016 in there
dfFinal["Reporting Year"].value_counts()

2021       1421
2020        582
2019        549
2016        549
2018        539
2017        539
6-2020       81
1-2020       81
5-2020       81
11-2020      81
7-2020       81
2-2021       81
12-2020      81
1-2021       81
2-2020       81
8-2020       81
10-2020      81
3-2020       81
9-2020       80
8-2021       80
12-2021      80
11-2021      80
9-2021       80
5-2021       80
10-2021      80
3-2021       80
7-2021       80
4-2021       80
6-2021       78
4-2020       77
Name: Reporting Year, dtype: int64

### Fix Port of Charleston, SC


1. The port name for Charleston, SC was misspelled as  "Port of Charleston, SC, SC" in the New data, so it was  changed to "Port of Charleston, SC"

The issues below were caused by adding the 2016 data from the Original Source to the Final dataframe<br>
2. For 2016, from the Original, the port name was : "Charleston, SC Port of", so I changed this to : "Port of Charleston, SC".<br>
3. For 2016, the Port ID was 773 and was changed to 775, to be consistent with the Port of Charleston Port ID in the new data. 

In [36]:
dfFinal["Port_Name"].value_counts()

Baltimore, MD                             173
Jacksonville, FL                          173
San Juan, PR                              173
Los Angeles, CA Port of                   173
Honolulu, O'ahu, HI                       173
New York, NY & NJ                         173
Savannah, GA Port of                      173
Long Beach, CA Port of                    173
Port Freeport, TX                         173
Virginia, VA, Port of                     173
Philadelphia Regional Port, PA            173
Mobile, AL                                173
Houston Port Authority, TX                173
Boston, MA                                172
Port Everglades, FL                       169
Alaska, AK Port of                        169
Wilmington, NC                            166
Port of Charleston, SC, SC                159
Wilmington, DE                            149
Tampa Port Authority, FL                  149
Tacoma, WA                                149
New Orleans, LA                   

#### First, Fix the Bad Spelling

In [37]:
dfFinal.shape

(6106, 11)

In [38]:
print(dfFinal.loc[dfFinal["Port_Name"].str.lower().str.contains("charles"),"Port_Name"].value_counts())

Port of Charleston, SC, SC          159
Lake Charles Harbor District, LA    103
Charleston, SC Port of               14
Name: Port_Name, dtype: int64


In [39]:
##  ACTUAL FIX
goodName= "Port of Charleston, SC"
badName= "Port of Charleston, SC, SC"

dfFinal["Port_Name"]=dfFinal["Port_Name"].str.replace(badName,goodName)

In [40]:
print(dfFinal.loc[dfFinal["Port_Name"].str.lower().str.contains("charles"),"Port_Name"].value_counts())
print(dfFinal.shape)

Port of Charleston, SC              159
Lake Charles Harbor District, LA    103
Charleston, SC Port of               14
Name: Port_Name, dtype: int64
(6106, 11)


#### Now, Change the Name from the 2016 Data to Match the New Name for Charleston

We added the Records for 2016 from the Original Data to the Final dataframe.  We need to make certain the Names for the Charleston, SC port match, and the Port ID.  Turns out we need to change both of these from the 2016 data to match the ones in the New Source!

In [41]:
#dfF = dfFinal.copy()
origName= "Charleston, SC Port of"
newName= "Port of Charleston, SC"
if dfFinal.loc[dfFinal["Port_Name"] == origName,'Port ID'].nunique() == 1:
    oldID=dfFinal.loc[dfFinal["Port_Name"] == origName,'Port ID'].values.tolist()[0]
    
if dfFinal.loc[dfFinal["Port_Name"] == newName,'Port ID'].nunique() == 1:
    newID = dfFinal.loc[dfFinal["Port_Name"] == newName,'Port ID'].values.tolist()[0]
    
# Change Port ID if needed
# I set this up for kind of a bulk use, jic, so check to make sure the Port IDs are different.  It probably 
# not a big deal, but never a bad idea to skip making a change if not needed. 
if newID != oldID:
    print("Change ID",oldID,newID)
    dfFinal.loc[dfFinal["Port_Name"] == origName,'Port ID'] = newID
    
# Change Original port name to the New port name
dfFinal.loc[dfFinal["Port_Name"] == origName,'Port_Name'] = newName

# Check Port ID to be certain it changed.
print(dfFinal.loc[dfFinal["Port_Name"] == newName,'Port ID'])
#dfFinal["Port_Name"]=dfFinal["Port_Name"].str.replace(origName,newName)

Change ID 773 775
124     775
125     775
126     775
127     775
683     775
684     775
685     775
686     775
687     775
2392    775
2393    775
2394    775
2395    775
2396    775
128     775
129     775
130     775
131     775
132     775
133     775
134     775
135     775
136     775
137     775
138     775
139     775
140     775
141     775
142     775
143     775
144     775
145     775
146     775
147     775
740     775
741     775
742     775
743     775
744     775
745     775
746     775
747     775
748     775
749     775
750     775
751     775
752     775
753     775
754     775
755     775
756     775
757     775
758     775
759     775
760     775
761     775
762     775
763     775
764     775
1752    775
1753    775
1754    775
1755    775
1756    775
1757    775
2038    775
2039    775
2040    775
2041    775
2042    775
2043    775
2415    775
2416    775
2417    775
2418    775
2419    775
2420    775
2421    775
2422    775
2423    775
2424    775
2425    77

In [42]:
# Check Port ID to be certain it changed.
print(dfFinal.loc[dfFinal["Port_Name"] == newName,'Port ID'].value_counts())
print(dfFinal.loc[dfFinal["Port_Name"] == newName,'Reporting Year'].value_counts())

775    173
Name: Port ID, dtype: int64
2021       31
2017       14
2020       14
2016       14
2018       14
2019       14
8-2021      3
1-2021      3
7-2021      3
10-2021     3
12-2020     3
1-2020      3
9-2020      3
3-2020      3
2-2021      3
11-2020     3
5-2021      3
5-2020      3
12-2021     3
10-2020     3
11-2021     3
6-2021      3
6-2020      3
4-2020      3
8-2020      3
2-2020      3
9-2021      3
7-2020      3
4-2021      3
3-2021      3
Name: Reporting Year, dtype: int64


In [43]:
# Sanity Check
print(dfFinal.loc[dfFinal["Port_Name"].str.lower().str.contains("charles"),"Port_Name"].value_counts())

Port of Charleston, SC              173
Lake Charles Harbor District, LA    103
Name: Port_Name, dtype: int64


In [45]:
#  OK, Final Sanity Inspection....
dfFinal["Port_Name"].value_counts().sort_index()

Alaska, AK Port of                        169
Baltimore, MD                             173
Beaumont, TX                              103
Boston, MA                                172
Cincinnati-Northern KY, Ports of           10
Cleveland-Cuyahoga Port, OH                68
Corpus Christi, TX                        105
Duluth-Superior, MN and WI                 81
Greater Baton Rouge, LA Port of           101
Gulfport, MS                               14
Honolulu, O'ahu, HI                       173
Houston Port Authority, TX                173
Huntington-Tristate, KY, OH, WV            10
Jacksonville, FL                          173
Kalama, WA Port of                         76
Lake Charles Harbor District, LA          103
Long Beach, CA Port of                    173
Longview, WA Port of                       80
Los Angeles, CA Port of                   173
Mid-America Port Commission                47
Mobile, AL                                173
New Bourbon Port, MO              

In [46]:
print(dfFinal.shape)
print(dfNew.shape)

(6106, 11)
(5557, 11)


###  Add Stations from Orig to New

The are 8 stations in the Orignal source that are not in the New Source.  We want to add these, but, we have to check a few things<br>
<ol>
    <li>Check to be certain the 8 stations do not exist under a different spelling in New source
    <li>If adding stations to Final dataframe from the Original, check Port ID to see if it needs to be changed
</ol>

In [47]:
# a = dfOldNo2016NoExtra["Port_Name"].unique().tolist()
# b = dfNewNoDwellNo2021["Port_Name"].unique().tolist()

a = dfOld["Port_Name"].unique().tolist()
b = dfNew["Port_Name"].unique().tolist()

In [48]:
stns2Get = set(a)-set(b)

In [49]:
#  OK, stations to add
stns2Get

{'Charleston, SC Port of',
 'Cincinnati-Northern KY, Ports of',
 'Gulfport, MS',
 'Huntington-Tristate, KY, OH, WV',
 'Illinois Waterway Ports and Terminals Port Statistical Area, IL',
 'Mid-Ohio Valley Port, OH and WV',
 'Portland, OR Port of',
 'St. Louis, MO and IL'}

####  Check Stations to Add Against Final Stations

So I am checking against the dfNew dataframe b/c I have already added data for 2016 from the Original source, so most of these stations would show in dfFinal for 2016.  So, it is best to use dfNew<br>
<p>
For this check, I am checking different substrings of the 8 stations to determine if these stations exist as a different spelling in the New source

In [50]:
## Check for stations in New to be certain we do not duplicate a station

print ('Check Charleston, SC Port of')
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("charles"),"Port_Name"].value_counts())

print("---------")
print ('Cincinnati-Northern KY, Ports of')
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("cincinanati"),"Port_Name"].value_counts())
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("norther"),"Port_Name"].value_counts())
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("ky"),"Port_Name"].value_counts())

print("---------")
print('Gulfport, MS')
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("gulf"),"Port_Name"].value_counts())

print("---------")
print('Huntington-Tristate, KY, OH, WV')
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("hunt"),"Port_Name"].value_counts())
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("tri"),"Port_Name"].value_counts())

print("---------")
print('Illinois Waterway Ports and Terminals Port Statistical Area, IL')
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("illinois"),"Port_Name"].value_counts())
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("term"),"Port_Name"].value_counts())
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("water"),"Port_Name"].value_counts())


print("---------")
print('Mid-Ohio Valley Port, OH and WV')
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("ohio"),"Port_Name"].value_counts())
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("valley"),"Port_Name"].value_counts())
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("mid"),"Port_Name"].value_counts())



print("---------")
print('Portland, OR Port of')
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("portland"),"Port_Name"].value_counts())

print("---------")
print('St. Louis, MO and IL')
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("lois"),"Port_Name"].value_counts())
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("st"),"Port_Name"].value_counts())
print(dfNew.loc[dfNew["Port_Name"].str.lower().str.contains("saint"),"Port_Name"].value_counts())



Check Charleston, SC Port of
Port of Charleston, SC, SC          159
Lake Charles Harbor District, LA     93
Name: Port_Name, dtype: int64
---------
Cincinnati-Northern KY, Ports of
Series([], Name: Port_Name, dtype: int64)
Northern Indiana, IN    61
Name: Port_Name, dtype: int64
Series([], Name: Port_Name, dtype: int64)
---------
Gulfport, MS
Series([], Name: Port_Name, dtype: int64)
---------
Huntington-Tristate, KY, OH, WV
Series([], Name: Port_Name, dtype: int64)
Port of Palm Beach District, FL           130
Plaquemines Port District, LA              95
Lake Charles Harbor District, LA           93
Southern Indiana Maritime District, IN     27
Name: Port_Name, dtype: int64
---------
Illinois Waterway Ports and Terminals Port Statistical Area, IL
Series([], Name: Port_Name, dtype: int64)
Series([], Name: Port_Name, dtype: int64)
Series([], Name: Port_Name, dtype: int64)
---------
Mid-Ohio Valley Port, OH and WV
Series([], Name: Port_Name, dtype: int64)
Series([], Name: Port_Name, dt

##### Summary

There were 2 hits from the list of 8 stations<br>
<ul>
    <li> Check Charleston, SC Port of
    <li> Portland, OR Port of
</ul>
<p>
 So, we need to removes these from the  list of stations to add

##### Remove Charleston, SC and Portland, OR from the list of Needed Stations

stns2Get is the list of stations in the Original data not found in the New Data.  However, the Port of Charleston, SC  IS in the New data, but it is misspelled, which will be fixed below.  So, we only need Port of Charleston data for 2016

In [51]:
stns2Get

{'Charleston, SC Port of',
 'Cincinnati-Northern KY, Ports of',
 'Gulfport, MS',
 'Huntington-Tristate, KY, OH, WV',
 'Illinois Waterway Ports and Terminals Port Statistical Area, IL',
 'Mid-Ohio Valley Port, OH and WV',
 'Portland, OR Port of',
 'St. Louis, MO and IL'}

In [52]:
# remove stations from list
stns2Get.remove('Charleston, SC Port of')
stns2Get.remove('Portland, OR Port of')
print(stns2Get)

{'St. Louis, MO and IL', 'Huntington-Tristate, KY, OH, WV', 'Gulfport, MS', 'Cincinnati-Northern KY, Ports of', 'Mid-Ohio Valley Port, OH and WV', 'Illinois Waterway Ports and Terminals Port Statistical Area, IL'}


#### Check Port IDs for Stations 

FOr the 6 stations left that we need to add from the Original source, we need to check the Port IDs to be certain there are no conflicts with existing stations

In [53]:
portIDs = list(dfOld.loc[dfOld["Port_Name"].isin(stns2Get),"Port ID"].unique())

In [54]:
portIDs

[2083, 2338, 2348, 2366, 2367, 7713]

In [55]:
dfFinal.loc[dfFinal["Port ID"].isin(portIDs)]

,Cargo Type,Port ID,Port_Name,Region,Reporting Year,State,Trade Type,Units,Port Ranking,Percent Change,Volume
284,CONTAINER,2083,"Gulfport, MS",Gulf Coast & Mississippi River,2016,Mississippi,EMPTY,Container TEUs,28.0,0,0
285,CONTAINER,2083,"Gulfport, MS",Gulf Coast & Mississippi River,2016,Mississippi,EXPORTS,Container TEUs,28.0,-8.00319,45290.9
286,CONTAINER,2083,"Gulfport, MS",Gulf Coast & Mississippi River,2016,Mississippi,IMPORTS,Container TEUs,28.0,20.3489,79431.5
287,CONTAINER,2083,"Gulfport, MS",Gulf Coast & Mississippi River,2016,Mississippi,TOTAL,Container TEUs,28.0,8.23595,124722
883,DRY BULK,2083,"Gulfport, MS",Gulf Coast & Mississippi River,2016,Mississippi,DOMESTIC,Short Tons,119.0,-37.94,176668
884,DRY BULK,2083,"Gulfport, MS",Gulf Coast & Mississippi River,2016,Mississippi,EXPORTS,Short Tons,119.0,0,0
885,DRY BULK,2083,"Gulfport, MS",Gulf Coast & Mississippi River,2016,Mississippi,FOREIGN,Short Tons,119.0,4.48322,554086
886,DRY BULK,2083,"Gulfport, MS",Gulf Coast & Mississippi River,2016,Mississippi,IMPORTS,Short Tons,119.0,4.48322,554086
887,DRY BULK,2083,"Gulfport, MS",Gulf Coast & Mississippi River,2016,Mississippi,TOTAL,Short Tons,119.0,-10.3352,730754
1053,DRY BULK,2338,"Cincinnati-Northern KY, Ports of",Gulf Coast & Mississippi River,2016,Kentucky and Ohio,DOMESTIC,Short Tons,4.0,-4.88136,3.8294e+07


Ok, no Port ID conflicts, so we can create a dataframe of data from the Original source for the 6 stations we want to add.  Note that since we already added 2016 data for all available stations, we need to skip 2016 so we do not create duplicates

In [56]:
#  Createa a dataframe of the 6 stations  for the Original source that we want to add to dfFinal.  
dfOldAdd2New = dfOld.loc[(dfOld["Port_Name"].isin(stns2Get)) & (dfOld["Reporting Year"] != '2016')]

In [57]:
## The old Sanity Checks...
print(dfOldAdd2New["Port ID"].unique())
print(dfOldAdd2New["Port_Name"].value_counts())
print(dfOldAdd2New["Reporting Year"].value_counts())

[2083 2338 2348 2366 2367 7713]
Gulfport, MS                                                       68
Cincinnati-Northern KY, Ports of                                   57
St. Louis, MO and IL                                               57
Huntington-Tristate, KY, OH, WV                                    51
Illinois Waterway Ports and Terminals Port Statistical Area, IL    27
Mid-Ohio Valley Port, OH and WV                                    26
Name: Port_Name, dtype: int64
2020    159
2017     44
2018     44
2019     39
Name: Reporting Year, dtype: int64


In [58]:
## One last Sanity Check before we combine
dfNew["Port ID"].loc[dfNew["Port ID"].isin(dfOldAdd2New["Port ID"].unique())]

Series([], Name: Port ID, dtype: int64)

### Add Original Stations Missing in New Source

In [59]:
dfFinal = pd.concat([dfFinal,dfOldAdd2New])

In [60]:
dfFinal.shape

(6392, 11)

### Remove Vessel Dwell Time Cargo Type

In [61]:
dfFinal["Cargo Type"].value_counts()

VESSEL DWELL TIMES             1927
TOTAL TONNAGE                  1445
DRY BULK                       1440
CONTAINER                       700
TOP 5 COMMODITIES               322
TOP 5 FOOD/FARM COMMODITIES     288
VESSEL CALLS                    270
Name: Cargo Type, dtype: int64

In [62]:
dfFinal.loc[dfFinal["Cargo Type"] == "VESSEL DWELL TIMES","Reporting Year"].value_counts()

6-2020     81
5-2020     81
7-2020     81
2-2021     81
11-2020    81
8-2020     81
2-2020     81
10-2020    81
12-2020    81
1-2020     81
3-2020     81
1-2021     81
3-2021     80
11-2021    80
9-2020     80
7-2021     80
9-2021     80
12-2021    80
10-2021    80
8-2021     80
5-2021     80
4-2021     80
6-2021     78
4-2020     77
Name: Reporting Year, dtype: int64

In [63]:
dfFinal = dfFinal.loc[dfFinal["Cargo Type"] != "VESSEL DWELL TIMES"]

In [64]:
dfFinal.shape

(4465, 11)

In [65]:
dfFinal["Reporting Year"].value_counts()

2021    1421
2020     741
2019     588
2017     583
2018     583
2016     549
Name: Reporting Year, dtype: int64

### Change Reporting Year to integer

New that the month-year dates have been removed with the dwell times, we can change the Reporting Year from String to integer

In [80]:
dfFinal["Reporting Year"]  = dfFinal["Reporting Year"] .astype(int)

### ADD 'TOP 5 COMODITITES','VESSEL CALL TIMES', 'TOP 5 FOOD/FARM COMMODITIES' for 2020 from the Original Source 

These Cargo Types are Important, but only exist in the Original Data for 2020 and in the New data for 2021.  So we need to get them out of the Original source for 2020.
<p>
    Steps Needed
    <ol>
        <li> There are some of these Cargo Types for the 6 ports already added from the Original Source, so we need to remove these Cargo Types from 2020 for these 6 ports
        <li> Create a Dataframe with only these 3 Cargo Types for 2020
        <li> Concat this dataframe with the final dataframe
        <li> Check and resolve ALL ports for Por ID conflicts caused by merging the Original Data and the New Data
    </ol>

#### For 2020, remove the 'TOP 5 COMODITITES','VESSEL CALL TIMES', 'TOP 5 FOOD/FARM COMMODITIES'

For 2020, the Cargo Types 'TOP 5 COMODITITES','VESSEL CALL TIMES', 'TOP 5 FOOD/FARM COMMODITIES' only exist for the 6 stations that were added from the Original Source.  For now, lets just remove these Cargo Types for 2020!  These will be left in 2021 because they exist for ALL stations

In [82]:
dfFinal.loc[(dfFinal["Reporting Year"] == 2020),"Cargo Type"].value_counts()

TOTAL TONNAGE                  265
DRY BULK                       265
CONTAINER                      116
TOP 5 COMMODITIES               36
VESSEL CALLS                    30
TOP 5 FOOD/FARM COMMODITIES     29
Name: Cargo Type, dtype: int64

This is a little tricky.  Remove all 2020 data first

In [83]:
dfFinalFinal = dfFinal.loc[dfFinal["Reporting Year"] != 2020]

In [84]:
print(dfFinal.shape)
print(dfFinalFinal.shape)

(4465, 11)
(3724, 11)


In [85]:
dfFinalFinal["Reporting Year"].value_counts()

2021    1421
2019     588
2017     583
2018     583
2016     549
Name: Reporting Year, dtype: int64

Now, create a dataframe that has  only 2020 data, but with the unwanted Cargo Types removed

In [86]:
dfFinal2020NoBadCargo = dfFinal.loc[(dfFinal["Reporting Year"] == 2020) & dfFinal["Cargo Type"].isin(["CONTAINER","TOTAL TONNAGE","DRY BULK"]) ]

In [89]:
# Check shape... the total record cound for 2020 is 741.  The new shape is 646... looks good
print(dfFinal.loc[dfFinal["Reporting Year"] == 2020].shape)
print(dfFinal2020NoBadCargo.shape)


(741, 11)
(646, 11)


Now, concat the new version of the 2020 data back into the finalfinal dataframe

In [90]:
dfFinalFinal = pd.concat([dfFinalFinal,dfFinal2020NoBadCargo])

OK, Lets sort the Final dataframe by Port Name, then Reporting Year to add a little sanity

In [91]:
dfFinalFinal.sort_values(by=["Port_Name","Reporting Year"],axis=0,inplace=True)

In [92]:
dfFinalFinal.shape

(4370, 11)

####  Get 2020 Data with only these Cargo Types:  'TOP 5 COMODITITES','VESSEL CALL TIMES', 'TOP 5 FOOD/FARM COMMODITIES' 

In [125]:
dfOld["Cargo Type"].value_counts()

TOTAL TONNAGE                  1150
DRY BULK                       1140
CONTAINER                       528
TOP 5 COMMODITIES               297
TOP 5 FOOD/FARM COMMODITIES     272
VESSEL CALLS                    250
Name: Cargo Type, dtype: int64

In [126]:
dfOld2020ExtraCargo = dfOld.loc[(dfOld["Reporting Year"] == '2020') & dfOld["Cargo Type"].isin(['TOP 5 COMMODITIES','VESSEL CALLS', 'TOP 5 FOOD/FARM COMMODITIES']) ]

In [128]:
dfOld2020ExtraCargo["Cargo Type"].value_counts()

TOP 5 COMMODITIES              297
TOP 5 FOOD/FARM COMMODITIES    272
VESSEL CALLS                   250
Name: Cargo Type, dtype: int64

In [129]:
dfOld2020ExtraCargo["Reporting Year"].value_counts()

2020    819
Name: Reporting Year, dtype: int64

In [130]:
dfOld2020ExtraCargo["Port_Name"].value_counts()

Wilmington, DE                                                     17
Boston, MA                                                         17
Tacoma, WA                                                         17
Port Freeport, TX                                                  17
Greater Baton Rouge, LA Port of                                    17
Cincinnati-Northern KY, Ports of                                   17
Long Beach, CA Port of                                             17
Texas City, TX                                                     17
Seattle, WA                                                        17
Lake Charles Harbor District, LA                                   17
Mid-America Port Commission                                        17
Los Angeles, CA Port of                                            17
Portland, OR Port of                                               17
Plaquemines Port District, LA                                      17
Longview, WA Port of

In [132]:
for portid in dfOld2020ExtraCargo["Port ID"].unique():
    a = list(dfOld2020ExtraCargo.loc[dfOld2020ExtraCargo["Port ID"] == portid,"Port_Name"].unique())
    print(portid,len(a),a)

149 1 ['Boston, MA']
398 1 ['New York, NY & NJ']
550 1 ['South Jersey Port, NJ']
552 1 ['Philadelphia Regional Port, PA']
554 1 ['Wilmington, DE']
700 1 ['Baltimore, MD']
766 1 ['Wilmington, NC']
773 1 ['Charleston, SC Port of']
776 1 ['Savannah, GA Port of']
1911 1 ['Port Everglades, FL']
1913 1 ['San Juan, PR']
2005 1 ['Mobile, AL']
2017 1 ['Jacksonville, FL']
2021 1 ['Tampa Port Authority, FL']
2031 1 ['Houston Port Authority, TX']
2083 1 ['Gulfport, MS']
2164 1 ['PortMiami, FL']
2248 1 ['Lake Charles Harbor District, LA']
2251 1 ['New Orleans, LA']
2252 1 ['Greater Baton Rouge, LA Port of']
2253 1 ['South Louisiana, LA, Port of']
2255 1 ['Plaquemines Port District, LA']
2306 1 ['Mid-America Port Commission']
2338 1 ['Cincinnati-Northern KY, Ports of']
2348 1 ['Huntington-Tristate, KY, OH, WV']
2351 1 ['New Bourbon Port, MO']
2358 1 ['Pittsburgh, PA Port of']
2366 1 ['Mid-Ohio Valley Port, OH and WV']
2367 1 ['St. Louis, MO and IL']
2393 1 ['Beaumont, TX']
2408 1 ['Port Freeport, TX

#### Add 2020 Cargo Types to Final Dataframe

In [136]:
dfFinalFinal = pd.concat([dfFinalFinal,dfOld2020ExtraCargo])

In [137]:
for portid in dfFinalFinal["Port ID"].unique():
    a = list(dfFinalFinal.loc[dfFinalFinal["Port ID"] == portid,"Port_Name"].unique())
    print(portid,len(a),a)

4820 1 ['Alaska, AK Port of']
700 1 ['Baltimore, MD']
2393 1 ['Beaumont, TX']
149 1 ['Boston, MA']
2338 1 ['Cincinnati-Northern KY, Ports of']
3217 1 ['Cleveland-Cuyahoga Port, OH']
2436 1 ['Corpus Christi, TX']
3924 1 ['Duluth-Superior, MN and WI']
2252 1 ['Greater Baton Rouge, LA Port of']
2083 1 ['Gulfport, MS']
4421 1 ["Honolulu, O'ahu, HI"]
2031 1 ['Houston Port Authority, TX']
2348 1 ['Huntington-Tristate, KY, OH, WV']
7713 1 ['Illinois Waterway Ports and Terminals Port Statistical Area, IL']
2017 1 ['Jacksonville, FL']
4626 1 ['Kalama, WA Port of']
2248 1 ['Lake Charles Harbor District, LA']
4110 1 ['Long Beach, CA Port of']
4622 1 ['Longview, WA Port of']
4120 1 ['Los Angeles, CA Port of']
2306 1 ['Mid-America Port Commission']
2366 1 ['Mid-Ohio Valley Port, OH and WV']
2032 1 ['Mobile, AL']
2351 1 ['New Bourbon Port, MO']
2251 1 ['New Orleans, LA']
398 1 ['New York, NY & NJ']
3743 1 ['Northern Indiana, IN']
4344 1 ['Oakland, CA Port of']
552 1 ['Philadelphia Regional Port, PA'

#### Check for Port Name Conflicts, Again

Because we added data from the Original Source again, we need to check for Port Name Conflicts, again...and resolve them... as a reminder, we had to fix Charleston, SC and Portland, OR already, and likely again

In [146]:
checkStrins = {}
checkStrins['Charleston, SC Port of'] = []
checkStrins['Charleston, SC Port of'].append('charles')

checkStrins['Cincinnati-Northern KY, Ports of'] = []
checkStrins['Cincinnati-Northern KY, Ports of'].append('cincinanati')
checkStrins['Cincinnati-Northern KY, Ports of'].append('north')
checkStrins['Cincinnati-Northern KY, Ports of'].append('ky')


checkStrins['Gulfport, MS'] = []
checkStrins['Gulfport, MS'].append("gulf")


checkStrins['Huntington-Tristate, KY, OH, WV'] = []
checkStrins['Huntington-Tristate, KY, OH, WV'].append("hunt")
checkStrins['Huntington-Tristate, KY, OH, WV'].append("tri")



checkStrins['Illinois Waterway Ports and Terminals Port Statistical Area, IL'] = []
checkStrins['Illinois Waterway Ports and Terminals Port Statistical Area, IL'].append("illinois")
checkStrins['Illinois Waterway Ports and Terminals Port Statistical Area, IL'].append("term")
checkStrins['Illinois Waterway Ports and Terminals Port Statistical Area, IL'].append("term")


checkStrins['Mid-Ohio Valley Port, OH and WV'] = []
checkStrins['Mid-Ohio Valley Port, OH and WV'].append("ohio")
checkStrins['Mid-Ohio Valley Port, OH and WV'].append("valley")
checkStrins['Mid-Ohio Valley Port, OH and WV'].append("mid")



checkStrins['Portland, OR Port of'] = []
checkStrins['Portland, OR Port of'].append("land")
checkStrins['Portland, OR Port of'].append("portland")


checkStrins['St. Louis, MO and IL'] = []
checkStrins['St. Louis, MO and IL'].append("louis")
checkStrins['St. Louis, MO and IL'].append('st')
checkStrins['St. Louis, MO and IL'].append('saint')



print ('Check Charleston, SC Port of')
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("charles"),"Port_Name"].value_counts())

print("---------")
print ('Cincinnati-Northern KY, Ports of')
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("cincinanati"),"Port_Name"].value_counts())
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("norther"),"Port_Name"].value_counts())
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("ky"),"Port_Name"].value_counts())

print("---------")
print('Gulfport, MS')
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("gulf"),"Port_Name"].value_counts())

print("---------")
print('Huntington-Tristate, KY, OH, WV')
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("hunt"),"Port_Name"].value_counts())
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("tri"),"Port_Name"].value_counts())

print("---------")
print('Illinois Waterway Ports and Terminals Port Statistical Area, IL')
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("illinois"),"Port_Name"].value_counts())
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("term"),"Port_Name"].value_counts())
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("water"),"Port_Name"].value_counts())


print("---------")
print('Mid-Ohio Valley Port, OH and WV')
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("ohio"),"Port_Name"].value_counts())
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("valley"),"Port_Name"].value_counts())
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("mid"),"Port_Name"].value_counts())



print("---------")
print('Portland, OR Port of')
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("portland"),"Port_Name"].value_counts())

print("---------")
print('St. Louis, MO and IL')
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("lois"),"Port_Name"].value_counts())
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("st"),"Port_Name"].value_counts())
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains("saint"),"Port_Name"].value_counts())


Check Charleston, SC Port of
Port of Charleston, SC              101
Lake Charles Harbor District, LA     96
Charleston, SC Port of               17
Name: Port_Name, dtype: int64
---------
Cincinnati-Northern KY, Ports of
Series([], Name: Port_Name, dtype: int64)
Northern Indiana, IN                86
Cincinnati-Northern KY, Ports of    67
Name: Port_Name, dtype: int64
Cincinnati-Northern KY, Ports of    67
Huntington-Tristate, KY, OH, WV     61
Name: Port_Name, dtype: int64
---------
Gulfport, MS
Gulfport, MS    82
Name: Port_Name, dtype: int64
---------
Huntington-Tristate, KY, OH, WV
Huntington-Tristate, KY, OH, WV    61
Name: Port_Name, dtype: int64
Plaquemines Port District, LA             98
Lake Charles Harbor District, LA          96
Port of Palm Beach District, FL           82
Huntington-Tristate, KY, OH, WV           61
Southern Indiana Maritime District, IN    27
Name: Port_Name, dtype: int64
---------
Illinois Waterway Ports and Terminals Port Statistical Area, IL
Illinois 

#### Fix Charleston and Portland Name Conflicts

In [150]:
goodName= "Port of Charleston, SC"
badName= "Charleston, SC Port of"
dfFinalFinal["Port_Name"]=dfFinalFinal["Port_Name"].str.replace(badName,goodName)

goodName= "Port of Portland, OR"
badName= "Portland, OR Port of"
dfFinalFinal["Port_Name"]=dfFinalFinal["Port_Name"].str.replace(badName,goodName)

In [151]:
for port in checkStrins:
    print(port)
    for substr in checkStrins[port]:
       print("   Substring:",substr)
       print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains(substr),"Port_Name"].value_counts())
       print("   --------")
    print("----------------------------------------------------------")
    print("----------------------------------------------------------")         

Charleston, SC Port of
   Substring: charles
Port of Charleston, SC              118
Lake Charles Harbor District, LA     96
Name: Port_Name, dtype: int64
   --------
----------------------------------------------------------
----------------------------------------------------------
Cincinnati-Northern KY, Ports of
   Substring: cincinanati
Series([], Name: Port_Name, dtype: int64)
   --------
   Substring: north
Northern Indiana, IN                86
Cincinnati-Northern KY, Ports of    67
Name: Port_Name, dtype: int64
   --------
   Substring: ky
Cincinnati-Northern KY, Ports of    67
Huntington-Tristate, KY, OH, WV     61
Name: Port_Name, dtype: int64
   --------
----------------------------------------------------------
----------------------------------------------------------
Gulfport, MS
   Substring: gulf
Gulfport, MS    82
Name: Port_Name, dtype: int64
   --------
----------------------------------------------------------
-------------------------------------------------------

In [153]:
substr="port"
print(dfFinalFinal.loc[dfFinalFinal["Port_Name"].str.lower().str.contains(substr),"Port_Name"].value_counts().sort_index())


Alaska, AK Port of                                                 118
Cincinnati-Northern KY, Ports of                                    67
Cleveland-Cuyahoga Port, OH                                         68
Greater Baton Rouge, LA Port of                                     94
Gulfport, MS                                                        82
Houston Port Authority, TX                                         118
Illinois Waterway Ports and Terminals Port Statistical Area, IL     27
Kalama, WA Port of                                                  93
Long Beach, CA Port of                                             118
Longview, WA Port of                                                97
Los Angeles, CA Port of                                            118
Mid-America Port Commission                                         64
Mid-Ohio Valley Port, OH and WV                                     26
New Bourbon Port, MO                                                50
Oaklan

### Check Port IDs for Final Dataframe

By adding the 2016 data from the Original Source, and teh 2020 data for some Cargo Types,  we have potentially created some Port ID mis-matches as these sometimes are changed by the Army Core of Engineers (?).  So we need to check Ports for multiple Port IDs

In [154]:
#  Get list of ports that have multiple port ids
ports=[]
for port in dfFinalFinal["Port_Name"].unique():
    tmp = dfFinalFinal.loc[dfFinalFinal["Port_Name"] == port]
    np=tmp["Port ID"].nunique()
    if np > 1:
       print(port)
       ports.append(port)
       print(tmp["Port ID"].unique())
       print("------------------------")
       print("")

Corpus Christi, TX
[2436 2423]
------------------------

Honolulu, O'ahu, HI
[4421 4420]
------------------------

Mobile, AL
[2032 2005]
------------------------

Port of Charleston, SC
[775 773]
------------------------

Port of Portland, OR
[4646 4644]
------------------------

PortMiami, FL
[1992 2164]
------------------------

Tacoma, WA
[4719 4720]
------------------------

Wilmington, DE
[555 554]
------------------------



Roh-Roh...  OK, Lets get the Port ID we need for each of these 6 ports

In [155]:
portIDsNeeded = {}
for port in ports:
    a = list(dfFinalFinal.loc[(dfFinalFinal["Port_Name"] == port) & (dfFinalFinal["Reporting Year"] == 2021),"Port ID"].unique())
    print(a,port)
    portIDsNeeded[port]= a[0]
    

[2436] Corpus Christi, TX
[4421] Honolulu, O'ahu, HI
[2032] Mobile, AL
[775] Port of Charleston, SC
[4646] Port of Portland, OR
[1992] PortMiami, FL
[4719] Tacoma, WA
[555] Wilmington, DE


In [156]:
portIDsNeeded

{'Corpus Christi, TX': 2436,
 "Honolulu, O'ahu, HI": 4421,
 'Mobile, AL': 2032,
 'Port of Charleston, SC': 775,
 'Port of Portland, OR': 4646,
 'PortMiami, FL': 1992,
 'Tacoma, WA': 4719,
 'Wilmington, DE': 555}

In [157]:
for port in ports:
    dfFinalFinal.loc[dfFinalFinal["Port_Name"] == port,'Port ID'] = portIDsNeeded[port]

In [158]:
#  Sanity Check
portIDs = {}
for port in dfFinalFinal["Port_Name"].unique():
    tmp = dfFinalFinal.loc[dfFinalFinal["Port_Name"] == port]
    np=tmp["Port ID"].nunique()
    if np > 1:
       print(port)
       
       print(tmp["Port ID"].unique())
       print("------------------------")
       print("")
    else:
       portIDs[port] = tmp['Port ID'].unique()[0]

In [159]:
for port,portID in sorted(portIDs.items()):
    print(portID,port)

4820 Alaska, AK Port of
700 Baltimore, MD
2393 Beaumont, TX
149 Boston, MA
2338 Cincinnati-Northern KY, Ports of
3217 Cleveland-Cuyahoga Port, OH
2436 Corpus Christi, TX
3924 Duluth-Superior, MN and WI
2252 Greater Baton Rouge, LA Port of
2083 Gulfport, MS
4421 Honolulu, O'ahu, HI
2031 Houston Port Authority, TX
2348 Huntington-Tristate, KY, OH, WV
7713 Illinois Waterway Ports and Terminals Port Statistical Area, IL
2017 Jacksonville, FL
4626 Kalama, WA Port of
2248 Lake Charles Harbor District, LA
4110 Long Beach, CA Port of
4622 Longview, WA Port of
4120 Los Angeles, CA Port of
2306 Mid-America Port Commission
2366 Mid-Ohio Valley Port, OH and WV
2032 Mobile, AL
2351 New Bourbon Port, MO
2251 New Orleans, LA
398 New York, NY & NJ
3743 Northern Indiana, IN
4344 Oakland, CA Port of
552 Philadelphia Regional Port, PA
2358 Pittsburgh, PA Port of
2255 Plaquemines Port District, LA
2416 Port Arthur, TX
1911 Port Everglades, FL
2408 Port Freeport, TX
775 Port of Charleston, SC
1983 Port of 

### Create an inventory to check the data

In [110]:
#dfFinal["Reporting Year"] = dfFinal["Reporting Year"].astype(int)
hist={}
for index,row in dfFinalFinal.iterrows():
    name = row["Port_Name"]
    year = row["Reporting Year"]               
    id = row["Port ID"]   
    if name not in hist:
        hist[name] = {}
    if year not in hist[name]:
        hist[name][year]=0
    hist[name][year]+=1

In [111]:
dates = list(dfFinal["Reporting Year"].unique())

In [112]:
dates

[2016, 2017, 2018, 2019, 2020, 2021]

In [113]:

porder = [2016,2017,2018,2019,2020,2021]
#fout = open("check.txt","w") 
# for year in porder:
#   fout.write(f" {year:7s}") 
# fout.write("\n")
# print("")
for port in sorted(hist.keys()):
 #   print(port,end=" ")
    for year in porder:
        if year not in hist[port]:
           hist[port][year] = " "
      
pd.DataFrame.from_dict(hist).T
        

,2016,2017,2018,2019,2020,2021
"Alaska, AK Port of",14,14,14,14,14,31
"Baltimore, MD",14,14,14,14,14,31
"Beaumont, TX",10,10,10,10,10,29
"Boston, MA",14,14,14,14,14,31
"Cincinnati-Northern KY, Ports of",10,10,10,10,10,
"Cleveland-Cuyahoga Port, OH",,10,10,10,10,28
"Corpus Christi, TX",10,10,10,10,10,31
"Duluth-Superior, MN and WI",10,10,10,10,10,31
"Greater Baton Rouge, LA Port of",10,10,10,10,10,27
"Gulfport, MS",14,14,14,9,14,


### Write Final File

In [160]:
dfFinalFinal.to_csv("portCatalogUpdateFinal2.csv",index=False)

In [116]:
dfCheck = pd.read_csv("portCatalogUpdateFinal2.csv")

In [117]:
dfCheck.shape

(4370, 11)

In [118]:
dfCheck["Reporting Year"].value_counts()

2021    1421
2020     646
2019     588
2018     583
2017     583
2016     549
Name: Reporting Year, dtype: int64

In [119]:
for year in dfCheck["Reporting Year"].unique():
    print(year)
    print(dfCheck.loc[dfCheck["Reporting Year"] == year,"Cargo Type"].value_counts())
    print("---------------------------------")

2016
TOTAL TONNAGE    225
DRY BULK         220
CONTAINER        104
Name: Cargo Type, dtype: int64
---------------------------------
2017
TOTAL TONNAGE    240
DRY BULK         235
CONTAINER        108
Name: Cargo Type, dtype: int64
---------------------------------
2018
TOTAL TONNAGE    240
DRY BULK         235
CONTAINER        108
Name: Cargo Type, dtype: int64
---------------------------------
2019
DRY BULK         245
TOTAL TONNAGE    235
CONTAINER        108
Name: Cargo Type, dtype: int64
---------------------------------
2020
TOTAL TONNAGE    265
DRY BULK         265
CONTAINER        116
Name: Cargo Type, dtype: int64
---------------------------------
2021
TOP 5 COMMODITIES              286
TOP 5 FOOD/FARM COMMODITIES    259
TOTAL TONNAGE                  240
DRY BULK                       240
VESSEL CALLS                   240
CONTAINER                      156
Name: Cargo Type, dtype: int64
---------------------------------


In [120]:
dfFinal["Port_Name"].value_counts().sort_index()

Alaska, AK Port of                                                 101
Baltimore, MD                                                      101
Beaumont, TX                                                        79
Boston, MA                                                         101
Cincinnati-Northern KY, Ports of                                    67
Cleveland-Cuyahoga Port, OH                                         68
Corpus Christi, TX                                                  81
Duluth-Superior, MN and WI                                          81
Greater Baton Rouge, LA Port of                                     77
Gulfport, MS                                                        82
Honolulu, O'ahu, HI                                                101
Houston Port Authority, TX                                         101
Huntington-Tristate, KY, OH, WV                                     61
Illinois Waterway Ports and Terminals Port Statistical Area, IL     27
Jackso

In [ ]:
dfFinal.loc[(dfFinal["Reporting Year"] == 2021) & (dfFinal["Cargo Type"] == "TOP 5 COMMODITIES")]

In [367]:
dfOld.loc[(dfOld["Cargo Type"] == "TOP 5 COMMODITIES"),"Cargo Type"].value_counts()

TOP 5 COMMODITIES    297
Name: Cargo Type, dtype: int64

In [368]:
dfFinal.loc[(dfFinal["Cargo Type"] == "TOP 5 COMMODITIES"),"Reporting Year"].value_counts()

2021    286
2020     36
Name: Reporting Year, dtype: int64

In [374]:
dfFinal.loc[(dfFinal["Reporting Year"] == 2020),"Cargo Type"].value_counts()

DRY BULK                       265
TOTAL TONNAGE                  265
CONTAINER                      116
TOP 5 COMMODITIES               36
VESSEL CALLS                    30
TOP 5 FOOD/FARM COMMODITIES     29
Name: Cargo Type, dtype: int64

In [379]:
dfFinal.loc[~((dfFinal["Cargo Type"].isin(["TOP 5 COMMODITIES","TOP 5 FOOD/FARM COMMODITIES","VESSEL CALLS"])) & (dfFinal["Reporting Year"] == 2020))].shape

(4370, 11)

In [380]:
dfFinal.shape

(4465, 11)

In [375]:
dfFinal.loc[~(dfFinal["Cargo Type"].isin(["DRY BULK","TOTAL TONNAGE","CONTAINER"])),"Cargo Type"].value_counts()

TOP 5 COMMODITIES              322
TOP 5 FOOD/FARM COMMODITIES    288
VESSEL CALLS                   270
Name: Cargo Type, dtype: int64

### Summary

1.  Add 2016 data from Original source.
1.  New Data, fix Charleston, SC Port Name
    The port name for Charleston, SC was misspelled as  "Port of Charleston, SC, SC", changed it to "Port of Charleston, SC"
    For 2016, from the Original, the port name was : "Charleston, SC Port of", so I changed this to : "Port of Charleston, SC, SC".  Also, the Port ID was 773 and was changed to 775, to be consistent with the Port of Charleston Port ID in the new data. 
    


-----------------------------------------------------------------------

## OLD

In [ ]:
dfNewNoDwellNo2021

In [ ]:
for year in dfNewNoDwellNo2021["Reporting Year"].unique().tolist():
    year=int(year)
    tmpOld = dfOldNo2016NoExtra.loc[dfOldNo2016NoExtra["Reporting Year"] == year]
    tmpNew = dfNewNoDwellNo2021.loc[dfNewNoDwellNo2021["Reporting Year"].astype(int) == year]
    string="Cargo Type"
    print("Year ",year)
#    print("Old\n",tmpOld[string].value_counts())
#    print("New\n",tmpNew[string].value_counts())
    
    for port in tmpNew["Port_Name"].unique().tolist():
        tO = tmpOld.loc[tmpOld["Port_Name"] == port]
        tN = tmpNew.loc[tmpNew["Port_Name"] == port]
        print(year,port,tO.shape[0],tN.shape[0])
   #     tN.drop("Report Year New",inplace=True,axis=1)
        tO.reset_index(inplace=True,drop=True)
        tN.reset_index(inplace=True,drop=True)
        
#         print(tO.columns)
#         print(tN.columns)
        tN["Reporting Year"] = tN["Reporting Year"].astype(int)
        col='Percent Change'
        tN[col]= tN[col].str.replace(",","")
        tN[col] = tN[col].astype(float)
        tO[col] = tO[col].apply(lambda x: '{:,.2f}'.format(x))
        tN[col] = tN[col].apply(lambda x: '{:,.2f}'.format(x))
        col='Volume'
        tN[col] = tN[col].str.replace(",","")
        tN[col] = tN[col].astype(float)
#         print("-------------------\nOLD")
#         print(tO)
#         print("-------------------\nNEW")
        
#         print(tN)
        
        # tO[col] = tO[col].apply(lambda x: '{:,.2f}'.format(x))
        # tN[col] = tN[col].apply(lambda x: '{:,.2f}'.format(x))
        
        if tO.shape[0] > 0:
            if len(tN.compare(tO)) > 0:
               print(tN.compare(tO))
        else:
            print("Mis-Match",port,tN.shape,tO.shape)
        

In [ ]:
print(dfNew.loc[dfNew["Reporting Year"].str[-4:].astype(int) == 2021,"Cargo Type"].value_counts())

In [ ]:
print(dfNew.loc[dfNew["Reporting Year"].str[-4:].astype(int) == 2021,"Cargo Type"].value_counts())

In [ ]:
print(dfOldNo2016.shape)
print(dfNewNoDwellNo2021.shape)

In [ ]:
dfNew.dtypes

In [ ]:
dfNew["Report Year New"] = dfNew["Reporting Year"].str[-4:]

In [ ]:
dfNew["Report Year New"].value_counts()

In [ ]:
dfOld["Reporting Year"].value_counts()

In [ ]:
print(len(dfOld["Port_Name"].value_counts()))
print(len(dfNew["Port_Name"].value_counts()))
dfNewNo2021 = dfNew.loc[dfNew["Report Year New"].astype(int) < 2021]
dfOldNo2016 = dfOld.loc[dfOld["Reporting Year"].astype(int) > 2016]

In [ ]:

print(dfNewNo2021["Report Year New"].value_counts())
print(dfOldNo2016["Reporting Year"].value_counts())

In [ ]:
for port in dfOld["Port_Name"].unique().tolist():
    tmpOld = dfOldNo2016.loc[dfOldNo2016["Port_Name"] == port]
    tmpNew = dfNewNo2021.loc[dfNewNo2021["Port_Name"] == port]
    diff = tmpNew.shape[0]-tmpOld.shape[0]
    print(diff,tmpOld.shape[0],tmpNew.shape[0])
    

In [ ]:
string="Cargo Type"
print(dfNewNo2021[string].value_counts())
print(dfOldNo2016[string].value_counts())

In [ ]:
for year in dfOldNo2016["Reporting Year"].unique().tolist():
    tmpOld = dfOldNo2016.loc[dfOldNo2016["Reporting Year"] == year]
    tmpNew = dfNewNo2021.loc[dfNewNo2021["Report Year New"].astype(int) == year]
    string="Cargo Type"
    print("Year ",year)
    print("Old\n",tmpOld[string].value_counts())
    print("New\n",tmpNew[string].value_counts())
    

In [ ]:

for year in dfOldNo2016["Reporting Year"].unique().tolist():
    tmpOld = dfOldNo2016.loc[dfOldNo2016["Reporting Year"] == year]
    tmpNew = dfNewNo2021.loc[dfNewNo2021["Report Year New"].astype(int) == year]
    string="Cargo Type"
    print("Year ",year)
#    print("Old\n",tmpOld[string].value_counts())
#    print("New\n",tmpNew[string].value_counts())
    
    for port in tmpOld["Port_Name"].unique().tolist():
        tO = tmpOld.loc[tmpOld["Port_Name"] == port]
        tN = tmpNew.loc[tmpNew["Port_Name"] == port]
        print(year,port,tO.shape[0],tN.shape[0])
        tN.drop("Report Year New",inplace=True,axis=1)
        tO.reset_index(inplace=True,drop=True)
        tN.reset_index(inplace=True,drop=True)
        
        print(tO.columns)
        print(tN.columns)
        tN["Reporting Year"] = tN["Reporting Year"].astype(int)
        col='Percent Change'
        tN[col] = tN[col].astype(float)
        tO[col] = tO[col].apply(lambda x: '{:,.2f}'.format(x))
        tN[col] = tN[col].apply(lambda x: '{:,.2f}'.format(x))
        col='Volume'
        tN[col] = tN[col].str.replace(",","")
        tN[col] = tN[col].astype(float)
#         print("-------------------\nOLD")
#         print(tO)
#         print("-------------------\nNEW")
        
#         print(tN)
        
        # tO[col] = tO[col].apply(lambda x: '{:,.2f}'.format(x))
        # tN[col] = tN[col].apply(lambda x: '{:,.2f}'.format(x))
        if tN.shape[0] > 0:
            print(tO.compare(tN))
        
    

tN.dtypes

tO.head()

tN.head()

